plugins/askrene/askrene.c

```diff
@@ -0,0 +1,923 @@
+			   tal_count(path->scidds));
+	if (num != tal_count(path->scidds)) {
+		const struct reserve *r = find_reserve(askrene->reserved, &path->scidds[num]);
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				    "Overflow reserving %zu: %s amount %s (%s reserved already)",
+				    num,
+				    fmt_short_channel_id_dir(tmpctx, &path->scidds[num]),
+				    fmt_amount_msat(tmpctx, path->amounts[num]),
+				    r ? fmt_amount_msat(tmpctx, r->amount) : "none");
+	}
+
+	response = jsonrpc_stream_success(cmd);
+	return command_finished(cmd, response);
+}
+
+static struct command_result *json_askrene_unreserve(struct command *cmd,
+						     const char *buffer,
+						     const jsmntok_t *params)
+{
+	struct reserve_path *path;
+	struct json_stream *response;
+	size_t num;
+	struct askrene *askrene = get_askrene(cmd->plugin);
+
+	if (!param(cmd, buffer, params,
+		   p_req("path", param_reserve_path, &path),
+		   NULL))
+		return command_param_failed();
+
+	num = reserves_remove(askrene->reserved, path->scidds, path->amounts,
+			      tal_count(path->scidds));
+	if (num != tal_count(path->scidds)) {
+		const struct reserve *r = find_reserve(askrene->reserved, &path->scidds[num]);
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				    "Underflow unreserving %zu: %s amount %s (%zu reserved, amount %s)",
+				    num,
+				    fmt_short_channel_id_dir(tmpctx, &path->scidds[num]),
+				    fmt_amount_msat(tmpctx, path->amounts[num]),
+				    r ? r->num_htlcs : 0,
+				    r ? fmt_amount_msat(tmpctx, r->amount) : "none");
+	}
+
+	response = jsonrpc_stream_success(cmd);
+	return command_finished(cmd, response);
+}
+
+static struct command_result *param_layername(struct command *cmd,
+					      const char *name,
+					      const char *buffer,
+					      const jsmntok_t *tok,
+					      const char **str)
+{
+	*str = tal_strndup(cmd, buffer + tok->start,
+			   tok->end - tok->start);
+	if (strstarts(*str, "auto."))
+		return command_fail_badparam(cmd, name, buffer, tok,
+					     "New layers cannot start with auto.");
+	return NULL;
+}
+
+static struct command_result *json_askrene_create_channel(struct command *cmd,
+							  const char *buffer,
+							  const jsmntok_t *params)
+{
+	const char *layername;
+	struct layer *layer;
+	const struct local_channel *lc;
+	struct node_id *n1, *n2;
+	struct short_channel_id *scid;
+	struct amount_msat *capacity;
+	struct json_stream *response;
+	int *direction;
+	struct amount_msat *htlc_min, *htlc_max, *base_fee;
+	u32 *proportional_fee;
+	u16 *delay;
+	struct askrene *askrene = get_askrene(cmd->plugin);
+
+	if (!param_check(cmd, buffer, params,
+			 p_req("layer", param_layername, &layername),
+			 p_req("node1", param_node_id, &n1),
+			 p_req("node2", param_node_id, &n2),
+			 p_req("short_channel_id", param_short_channel_id, &scid),
+			 p_req("capacity_msat", param_msat, &capacity),
+			 p_req("direction", param_zero_or_one, &direction),
```

Comment:
As a suggestion, why not deduce the direction from the ordering of the nodes in the input?
For example, I want to add some information about the channel from node A to B, I don't have to worry
which direction is that I could just say: `create-channel layer A B ...`, and the plugin can deduce the direction
asking `dir = A<B ? 0 : 1` (I know there is some helper to do just that, but I don't remember right now the name).

Reply:
Hmm, yes, that's nicer.  Rename node1/node2 to source and destination.  Will fix...